### DISCUSSION TITLE
Custom/intermediate data (ctx.session?)  in backtest result

### DISCUSSION THREAD

**Discussion by SW4T400** [edtechre/pybroker#135] [Q&A] [answered]
I will use your [latest example](https://medium.com/@edtechre/portfolio-optimization-with-pybroker-5ce0af389bfa) for my question:

```
def rebalance(ctxs: dict[str, ExecContext]):
     if start_of_month(ctxs):
        target = 1 / len(ctxs)  # Equal weighting
        set_target_shares(ctxs, {symbol: target for symbol in ctxs.keys()})
```

```
strategy = Strategy(YFinance(), start_date='1/1/2018', end_date='1/1/2023')
strategy.add_execution(None, ['TSLA', 'NFLX', 'AAPL', 'NVDA', 'AMZN'])
strategy.set_after_exec(rebalance)
result = strategy.backtest()
```

The ACTUAL shares that pybroker was able to buy will be visible in result.orders & result.positions.
from there one can calculate the ACTUAL WEIGHTS that were achieved.

BUT  I might have additional logic in there (Order size limits, a no trade zone if actual and target weights are too similar etc.)
and I woul also like to track the TARGET WEIGHTS or some other custom_data for that matter.

1) I always have access to the global ``pybroker.param`` and it appears it can take dataframes - should I really write a timeseries dataframe to this?

2) But I also saw the [``ctx.session`` ](https://www.pybroker.com/en/dev/notebooks/FAQs.html#...-persist-data-across-bars?)
Is there an option to output session data? @edtechre 

I am currently trying to implement my [favorites authors trading system](https://github.com/robcarver17/pysystemtrade/tree/master)
in pybroker and find myself "wanting" to output some intermediate calculation results aka custom_data.

Any thoughts or even solutions?

**Comment by edtechre** [comment, marked as answer]
Hi @SW4T400,

PyBroker does not have an option for outputting session data. I think the approach in 1) is the best way to achieve this. I would recommend doing this by appending to a list of tuples, and then creating the DataFrame afterwards:

`df = pd.DataFrame(pybroker.param('data'), columns=columns)`
